posthog/models/integration.py

```diff
@@ -33,6 +46,127 @@ class IntegrationKind(models.TextChoices):
+                token_url="https://slack.com/api/oauth.v2.access",
+                client_id=from_settings["SLACK_APP_CLIENT_ID"],
+                client_secret=from_settings["SLACK_APP_CLIENT_SECRET"],
+                scope="channels:read,groups:read,chat:write,chat:write.customize",
+                id_path="team.id",
+            )
+        elif kind == "salesforce":
+            if not settings.SALESFORCE_CONSUMER_KEY or not settings.SALESFORCE_CONSUMER_SECRET:
+                raise NotImplementedError("Salesforce app not configured")
+
+            return OauthConfig(
+                authorize_url="https://login.salesforce.com/services/oauth2/authorize",
+                token_url="https://login.salesforce.com/services/oauth2/token",
+                client_id=settings.SALESFORCE_CONSUMER_KEY,
+                client_secret=settings.SALESFORCE_CONSUMER_SECRET,
+                scope="full",
+                id_path="instance_url",
+            )
+
+        raise NotImplementedError(f"Oauth config for kind {kind} not implemented")
+
+    @classmethod
+    def redirect_uri(cls, kind: str, next: str = "") -> str:
+        # The redirect uri is fixed but should always be https and include the "next" parameter for the frontend to redirect
+        return f"{settings.SITE_URL.replace('http://', 'https://')}/integrations/{kind}/callback?next={next}"
+
+    @classmethod
+    def authorize_url(cls, kind: str, next: str = "") -> str:
+        oauth_config = cls.oauth_config_for_kind(kind)
+
+        return f"{oauth_config.authorize_url}?client_id={oauth_config.client_id}&scope={oauth_config.scope}&redirect_uri={cls.redirect_uri(kind, next)}"
+
+    @classmethod
+    def integration_from_oauth_response(
+        cls, kind: str, team_id: str, created_by: User, params: dict[str, str]
+    ) -> Integration:
+        oauth_config = cls.oauth_config_for_kind(kind)
+
+        next = params["next"]
+        code = params["code"]
+
+        res = requests.post(
+            oauth_config.token_url,
+            data={
+                "client_id": oauth_config.client_id,
+                "client_secret": oauth_config.client_secret,
+                "code": code,
+                "redirect_uri": OauthIntegration.redirect_uri(kind, next=next),
+                "grant_type": "authorization_code",
+            },
+        )
+
+        config: dict = res.json()
+
+        if res.status_code != 200 or not config.get("access_token"):
+            raise Exception("Oauth error")
+
+        integration_id: Any = config
+
+        for key in oauth_config.id_path.split("."):
+            integration_id = integration_id.get(key)
+
+        if not isinstance(integration_id, str):
+            raise Exception("Oauth error")
+
+        sensitive_config: dict = {
+            "access_token": config.pop("access_token"),
+            "refresh_token": config.pop("refresh_token", None),
+            "id_token": config.pop("id_token", None),
+        }
```

Comment:
Should we do something with the `refresh_token`? From what I remember, OAuth has a mode (the default?) where you need to refresh the access token whenever it's close to expiry. Doing that for users automatically is for me the biggest win we can have by doing "oauth" natively instead of via Hog. The way I read the code now, if you set up an OAuth integration, it'll just stop working after a day/week/month when the token expires.

Reply:
There isn't actually a refresh token. I just added as a sanity to make sure we pull it out. and keep it in the "sensitive" field. 

With both slack and salesforce they return long lived tokens (and a bunch of others that I tests with as well) which seems to be common for this kind of delegated oauth flow (as opposed to a typical user login flow).

So yeah - when we have a need for it, then we implement it.

Reply:
How long is long lived?

Reply:
Added a comment to that effect